src/libsystemd/sd-event/sd-event.c

```diff
@@ -3977,3 +4031,96 @@ _public_ int sd_event_source_set_floating(sd_event_source *s, int b) {
 
         return 1;
 }
+
+static int ratelimit_timeout_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        int r;
+
+        sd_event_source *source = (sd_event_source *) userdata;
+
+        source->ratelimit->effective = false;
+
+        if (source->enabled != SD_EVENT_OFF) {
+                r = source_enable(source, source->enabled);
+                if (r < 0)
+                        return r;
+        }
+
+        r = sd_event_source_set_enabled(s, SD_EVENT_OFF);
+        if (r < 0)
+                return r;
+
+        return 1;
+}
+
+static int source_disable_ratelimit(sd_event_source *s) {
+        assert(s);
+
+        if (!s->ratelimit)
+                return 0;
+
+        if (s->ratelimit->effective && s->enabled != SD_EVENT_OFF) {
+                int r;
+
+                r = source_enable(s, s->enabled);
+                if (r < 0)
+                        return r;
+        }
+
+        s->ratelimit->timeout_source = sd_event_source_unref(s->ratelimit->timeout_source);
+        s->ratelimit = mfree(s->ratelimit);
+
+        return 1;
+}
+
+_public_ int sd_event_source_set_ratelimit(sd_event_source *s, uint64_t interval_usec, uint64_t burst) {
+        int r;
+
+        assert_return(s, -EINVAL);
+
+        r = source_disable_ratelimit(s);
+        if (r < 0)
+                return r;
```

Comment:
hmm, i#d prefer if this function would be atomic in behaviour, i.e. fail and make no changes or suceed and make the requested changes. Right now, if the new0() below fails we'll have disabled the ratelimit, but not enabled the new one.

i.e. allocate everything into temporary variables, then install them once you acquired them all, and don#t change state before that completed